### PR TITLE
server: support command get meta.

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut  @disksing

### CODE REVIEW

File: server/cluster.go [hunk 1]
```diff
@@ -43,9 +45,15 @@ var (
 type raftCluster struct {
 	s *Server
 
-	meta        protopb.Cluster
 	clusterRoot string
 
+	// For following cache meta.
```
Comment (ngaut) [inline server/cluster.go]: Use rust style mutex, Please embed protected fields to an unnamed structure.

**Comment by siddontang** [comment]
PTAL

File: server/cluster.go [hunk 2]
```diff
@@ -276,12 +285,111 @@ func (s *Server) bootstrapCluster(clusterID uint64, req *protopb.BootstrapReques
 	}
 
 	c := s.newCluster(clusterID, clusterMeta)
-	c.nodes[req.GetNode().GetNodeId()] = *req.GetNode()
+
+	c.mu.Lock()
```
Comment (ngaut) [inline server/cluster.go]: mu := &c.mu, then use mu instead of c.mu

File: server/cluster.go [hunk 3]
```diff
@@ -276,12 +285,111 @@ func (s *Server) bootstrapCluster(clusterID uint64, req *protopb.BootstrapReques
...
 
 	return nil
 }
+
+func (c *raftCluster) GetNode(nodeID uint64) (*protopb.Node, error) {
+	if nodeID == 0 {
+		return nil, errors.Errorf("invalid zero node id")
+	}
+
+	c.mu.RLock()
```
Comment (ngaut) [inline server/cluster.go]: Ditto.

File: server/cluster.go [hunk 4]
```diff
@@ -276,12 +285,111 @@ func (s *Server) bootstrapCluster(clusterID uint64, req *protopb.BootstrapReques
...
+
+	return &node, nil
+}
+
+func (c *raftCluster) GetStore(storeID uint64) (*protopb.Store, error) {
+	if storeID == 0 {
+		return nil, errors.Errorf("invalid zero store id")
+	}
+
+	c.mu.RLock()
```
Comment (ngaut) [inline server/cluster.go]: Ditto.

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @disksing

File: server/util.go [hunk 5]
```diff
@@ -30,6 +36,21 @@ func getValue(c *clientv3.Client, key string) ([]byte, error) {
...
+// return boolean to indicate whether the key exists or not.
+// TODO: return the value revision for outer use.
+func getProtoMsg(c *clientv3.Client, key string, msg proto.Message, opts ...clientv3.OpOption) (bool, error) {
+	value, err := getValue(c, key, opts...)
+	if err != nil || value == nil {
+		return false, errors.Trace(err)
+	}
+
+	if err = proto.Unmarshal(value, msg); err != nil {
+		return false, errors.Trace(err)
```
Comment (disksing) [inline server/util.go]: Should return `true`? Since the key DOES exist.

Comment (siddontang) [inline server/util.go]: here it is for error, so I think either is ok because we don't care it outer.

Comment (siddontang) [inline server/util.go]: Btw, I will use revision instead of this boolean later.

File: server/cluster.go [hunk 6]
```diff
@@ -276,12 +286,114 @@ func (s *Server) bootstrapCluster(clusterID uint64, req *protopb.BootstrapReques
...
+
+	// Etcd range search is for range [searchKey, endKey), if we want to get
+	// the latest region, we should use next max end key for search range.
+	// TODO: we can generate these keys in initialization.
+	maxEndKey := makeRegionSearchKey(c.clusterRoot, []byte{})
+	maxSearchEndKey := maxEndKey + "\x00"
+
+	// Find the first region with end key >= searchKey
+	region := protopb.Region{}
+	ok, err := getProtoMsg(c.s.client, searchKey, &region, clientv3.WithRange(string(maxSearchEndKey)), clientv3.WithLimit(1))
```
Comment (disksing) [inline server/cluster.go]: Why not cache regions?

Comment (siddontang) [inline server/cluster.go]: It is not easy to cache all regions, after we implement base pd, we will try to support it, so not now.

**Comment by disksing** [comment]
LGTM

**Pull request merged**